actions/actions_whitebox_test.go

```diff
@@ -0,0 +1,184 @@
+		fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns] = []interface{}{"bcid0", "bcid1"}
+		changes, err := fxt.WorkItems[0].ChangeSet(wiCopy)
+		require.NoError(t, err)
+		require.Len(t, changes, 1)
+		require.Equal(t, workitem.SystemState, changes[0].AttributeName)
+		require.Equal(t, "open", changes[0].NewValue)
+		require.Equal(t, "new", changes[0].OldValue)
+	})
+
+	s.T().Run("column changes", func(t *testing.T) {
+		wiCopy := createWICopy(*fxt.WorkItems[0], "new", []interface{}{"bcid0"})
+		fxt.WorkItems[0].Fields[workitem.SystemState] = "new"
+		fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns] = []interface{}{"bcid0", "bcid1"}
+		changes, err := fxt.WorkItems[0].ChangeSet(wiCopy)
+		require.NoError(t, err)
+		require.Len(t, changes, 1)
+		fmt.Println(changes[0].OldValue)
+		fmt.Println(changes[0].NewValue)
+		require.Equal(t, workitem.SystemBoardcolumns, changes[0].AttributeName)
+		require.Equal(t, wiCopy.Fields[workitem.SystemBoardcolumns], changes[0].OldValue)
+		require.Equal(t, fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns], changes[0].NewValue)
+	})
+
+	s.T().Run("multiple changes", func(t *testing.T) {
+		wiCopy := createWICopy(*fxt.WorkItems[0], "open", []interface{}{"bcid0"})
+		fxt.WorkItems[0].Fields[workitem.SystemState] = "new"
+		fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns] = []interface{}{"bcid0", "bcid1"}
+		changes, err := fxt.WorkItems[0].ChangeSet(wiCopy)
+		require.NoError(t, err)
+		require.Len(t, changes, 2)
+		// we intentionally test the order here as the code under test needs
+		// to be expanded later, supporting more changes and this is an
+		// integrity test on the current impl.
+		require.Equal(t, workitem.SystemState, changes[0].AttributeName)
+		require.Equal(t, "new", changes[0].NewValue)
+		require.Equal(t, "open", changes[0].OldValue)
+		require.Equal(t, workitem.SystemBoardcolumns, changes[1].AttributeName)
+		require.Equal(t, fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns], changes[1].NewValue)
+		require.Equal(t, wiCopy.Fields[workitem.SystemBoardcolumns], changes[1].OldValue)
+	})
+
+	s.T().Run("new instance", func(t *testing.T) {
+		fxt.WorkItems[0].Fields[workitem.SystemState] = "new"
+		fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns] = []interface{}{}
+		changes, err := fxt.WorkItems[0].ChangeSet(nil)
+		require.NoError(t, err)
+		require.Len(t, changes, 1)
+		require.Equal(t, workitem.SystemState, changes[0].AttributeName)
+		require.Equal(t, "new", changes[0].NewValue)
+		require.Nil(t, changes[0].OldValue)
+	})
+}
+
+func (s *ActionSuite) TestActionExecution() {
+	fxt := tf.NewTestFixture(s.T(), s.DB, tf.CreateWorkItemEnvironment(), tf.WorkItems(2))
+	userID := fxt.Identities[0].ID
+
+	s.T().Run("by Old New", func(t *testing.T) {
+		fxt.WorkItems[0].Fields[workitem.SystemState] = "new"
+		fxt.WorkItems[0].Fields[workitem.SystemBoardcolumns] = []interface{}{"bcid0", "bcid1"}
+		newVersion := createWICopy(*fxt.WorkItems[0], "open", []interface{}{"bcid0", "bcid1"})
```

Comment:
A test does not need to be future proof. It is supposed to fail early. This might be nice, but not necessary.

Reply:
Remember that I didn't use `WorkItemTypeRepository.CreateFromModel` in the test fixture? Do you recall how long it took you to search for the issue and how long it took me to find the cause? 

It was a waste of time if I had been using `CreateFromModel` from day 1.

It is the same with this piece of code. When a test fails I wouldn't know if your copy structure is wrong because it looks about right. But if there's a new field in the work item structure I most likely won't see that I have to adjust your copy mechanism and I sure wouldn't want to. 

This comment still holds true:

> And a function that is called createWICopy but sets some fields seems just wrong.

A while back the code was full of these helper functions and every piece in the controller package used them which made it really hard to clean up. I'm not allowing that we introduce those useless helper functions again.

Reply:
This is a helper function *in a test*. This is not even part of the WIT code. If I would have put this directly in the test code, no one would notice. This is just ridiculous. We need to stop nitpicking on this detail level.

Reply:
And the real issues not even get addresses. Like the use of constants all over the tests. *Those are the real issues* that will break the system and no test will ever notice because of "code beauty".